File: forge-web/lib/schemas/guideline.ts
import { z } from "zod";

const guidelineWeight = z.enum(["must", "should", "may"]);

export const guidelineCreateSchema = z.object({
  title: z.string().min(1, "Title is required"),
  scope: z.string().min(1, "Scope is required"),
  content: z.string().min(1, "Content is required"),
  rationale: z.string().optional(),
  examples: z.array(z.string()).optional(),
  tags: z.array(z.string()).optional(),
  weight: guidelineWeight.default("should"),
});

export const guidelineUpdateSchema = z.object({
  title: z.string().min(1).optional(),
  content: z.string().optional(),
  status: z.enum(["ACTIVE", "DEPRECATED"]).optional(),
  rationale: z.string().optional(),
  scope: z.string().optional(),
  examples: z.array(z.string()).optional(),
  tags: z.array(z.string()).optional(),
  weight: guidelineWeight.optional(),
  derived_from: z.string().optional(),
});

export type GuidelineCreateForm = z.infer<typeof guidelineCreateSchema>;
export type GuidelineUpdateForm = z.infer<typeof guidelineUpdateSchema>;

File: platform/app/models/guideline.py
from sqlalchemy import String, Text, Integer, ForeignKey, CheckConstraint
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models.base import TimestampMixin


class Guideline(Base, TimestampMixin):
    __tablename__ = "guidelines"
    __table_args__ = (
        CheckConstraint("weight IN ('must', 'should', 'may')", name="valid_weight"),
        CheckConstraint("status IN ('ACTIVE', 'DEPRECATED')", name="valid_guideline_status"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    project_id: Mapped[int | None] = mapped_column(ForeignKey("projects.id"))
    external_id: Mapped[str] = mapped_column(String(20), nullable=False)
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    scope: Mapped[str] = mapped_column(String(50), nullable=False, server_default="general")
    content: Mapped[str] = mapped_column(Text, nullable=False)
    rationale: Mapped[str | None] = mapped_column(Text)
    weight: Mapped[str] = mapped_column(String(10), nullable=False, server_default="should")
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="ACTIVE")
    examples: Mapped[list[str] | None] = mapped_column(ARRAY(String))

    project: Mapped["Project | None"] = relationship(back_populates="guidelines")
